enhancements/etcd/storage-migration-for-etcd-encryption.md

```diff
@@ -0,0 +1,122 @@
+---
+title: Neat-Enhancement-Idea
+authors:
+  - "@sanchezl"
+reviewers:
+  - TBD
+  - "@deads2k"
+  - "@enj"
+approvers:
+  - TBD
+creation-date: 2019-09-11
+last-updated: 2019-09-11
+status: provisional|implementable|implemented|deferred|rejected|withdrawn|replaced
+---
+
+# storage-migration-for-etcd-encryption
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift/docs]
+
+## Summary
+
+Introduce the `openshift-storage-migration-operator` to install the `StorageVersionMigration` CRD and the corresponding  `kube-storage-version-migrator` controller.  `StorageVersionMigration` resources can then be created by components, (initially  `kube-apiserver` and `openshift-apiserver`) to kick off storage migrations for resources.
+
+The first known usage is for etcd encryption at rest: `secrets` (`kube-apiserver`) and `oauthaccesstokens` (`openshift-apiserver`).
+
+## Motivation
+
+The immediate need for storage migration is to support the enablement of etcd data storage encryption. Encrypting etcd data storage involves rotating encryption keys periodically, which requires that stored data be migrated after choosing a new write key and before removing an old read key so that api servers can discard old encryption keys periodically and not lose the ability to decode the pre-existing data.
+
+We will use the upstream `kube-storage-version-migrator` (the controller-based successor to `oc adm migrate`). While upstream usage of the `kube-storage-version-migrator` is focused on migrating to new storage versions, we should be able to explicitly trigger the `kube-storage-version-migrator` when we need to migrate the encryption key used for storage, even when the storage version itself has not changed.
```

Comment:
do we need an operator running the migrator? Or will the operator be the migrator? I don't see a reason for added complexity splitting those two.

Reply:
Keeping the two separate will allow us to deploy the upstream controller as-is and using its native configuration.

While we don't have a lot of operational knowledge to embed in the operator at the moment (other than here are the namespaces, role bindings, deployments, etc.. needed to run the controller), it does give up a place to add new operational behavior in the future.

Reply:
sgtm.